File: src/app/flutter/convention_list/lib/models/convention.dart
import 'package:convention_list/models/position.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import 'category.dart';
import 'converters/category_converter.dart';

part 'convention.freezed.dart';
part 'convention.g.dart';

@freezed
class Convention with _$Convention {
  const Convention._();

  const factory Convention({
    required String id,
    required String name,
    required DateTime startDate,
    required DateTime endDate,
    required bool isApproved,
    String? city,
    String? country,
    String? postalCode,
    Position? position,
    String? description,
    String? websiteAddress,
    String? venueName,
    String? address1,
    String? address2,
    String? state,
    @CategoryConverter() @Default(Category.unlisted) Category category,
  }) = _Convention;

  factory Convention.fromJson(Map<String, dynamic> json) => _$ConventionFromJson(json);

  String getLocationInfo() {
    String info = '$city';
    if (state != null) {
      info += ', $state, $country';
    } else {
      info += ', $country';
    }
    return info;
  }
}

File: src/api/ConventionList.Core/Models/Convention.cs
using System.ComponentModel.DataAnnotations;
using NetTopologySuite.Geometries;

namespace ConventionList.Core.Models;

public class Convention
{
    public Guid Id { get; set; }

    public string? SubmitterId { get; set; }

    public Category? Category { get; set; }

    public string? ExternalId { get; set; }

    public string? ExternalSource { get; set; }

    [Required]
    public string Name { get; set; } = "";

    public string? Description { get; set; }

    [Required]
    public DateTime StartDate { get; set; }

    [Required]
    public DateTime EndDate { get; set; }

    public string? WebsiteAddress { get; set; }

    public string? VenueName { get; set; }

    public string? Address1 { get; set; }

    public string? Address2 { get; set; }

    public string? City { get; set; }

    public string? State { get; set; }

    public string? PostalCode { get; set; }

    public string? Country { get; set; } = "US";

    public Point? Position { get; set; }

    public bool IsApproved { get; set; }

    public string Editor { get; set; } = "";
}
